component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -134,16 +138,44 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
     private void validateLocalUser(OAuthTokenReqMessageContext tokReqMsgCtx, Map<String, String> requestParams)
             throws IdentityOAuth2Exception {
 
+        String userId = null;
+        try {
+            userId = tokReqMsgCtx.getAuthorizedUser().getUserId();
+        } catch (UserIdNotFoundException e) {
+            handleException(OAuth2ErrorCodes.SERVER_ERROR, e);
+        }
+
         AbstractUserStoreManager userStoreManager = TokenExchangeUtils.getUserStoreManager(tokReqMsgCtx);
         for (UserOperationEventListener listener : TokenExchangeServiceComponent.getUserOperationEventListeners()) {
             try {
                 listener.doPostAuthenticate(tokReqMsgCtx.getAuthorizedUser().getUserName(), false,
                         userStoreManager);
             } catch (UserStoreException e) {
+                if (e.getCause() instanceof AccountLockException) {
+                    String errorMessage = "Local user authorization failed: linked local account with id " + userId +
+                            " is locked";
+
+                    if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                        DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder =
+                                new DiagnosticLog.DiagnosticLogBuilder(
+                                        Constants.LogConstants.COMPONENT_ID,
+                                        Constants.LogConstants.ActionIDs.AUTHORIZE_LINKED_LOCAL_USER
+                                );
+                        diagnosticLogBuilder
+                                .resultMessage(errorMessage)
+                                .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                                .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                        LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
```

Comment:
are we logging only the account locked case?

Reply:
added a log to handle other authorization failures as well